simexpal/base.py

```diff
@@ -218,10 +223,20 @@ def revisions_for_experiment(exp_info):
 				yield from itertools.product([exp_info], revisions_for_experiment(exp_info),
 						selection.variations)
 
-		key = lambda x: (x[0].name, x[1].name if x[1] is not None else '_none',
-						 [sub_var.name for sub_var in x[2]])
-		for experiment_info, revision, variation in self._expand_matrix(extract, key=key):
-			yield Experiment(self, experiment_info, revision, variation)
+		if self._runs_discovered:
+			yield from self._runs.keys()
+
+		elif self._experiments_discovered:
+			yield from self._experiments.values()
```

Comment:
This is strange. I would expect that `all_experiments` only deals with experiments, not runs. Refactor this so that:
* If `not self._experiments_discovered`, some internal helper function is called to discover the experiments.
* Use the same helper function in `discover_all_runs()`. Note that that there can be experiments without runs but not the other way around. Hence, once the `_experiments` dict is built, `discover_all_runs()` can simply get its experiments from that dict (inserting into it is never necessary).

Reply:
> There could be a case where discover_all_runs() was already called.

True, but then `_experiments` should be filled already. `discover_all_runs()` and `all_experiments()` should both call a (single) "fill `_experiments`" helper function.

Reply:
done